File: maximal_square.rb
# Maximal Square: Largest square of 1s in matrix

def maximal_square(matrix)
  return 0 if matrix.empty?
  
  rows, cols = matrix.length, matrix[0].length
  dp = Array.new(rows) { Array.new(cols, 0) }
  max_side = 0
  
  rows.times do |i|
    cols.times do |j|
      if matrix[i][j] == '1'
        if i == 0 || j == 0
          dp[i][j] = 1
        else
          dp[i][j] = [dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]].min + 1
        end
        max_side = dp[i][j] if dp[i][j] > max_side
      end
    end
  end
  
  max_side * max_side
end

matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
puts maximal_square(matrix)

File: maximal_square.js
// Maximal Square: Find largest square submatrix containing only 1s
function maximalSquare(matrix) {
    if (!matrix.length || !matrix[0].length) return 0;
    let m = matrix.length, n = matrix[0].length;
    let dp = Array(m).fill(0).map(() => Array(n).fill(0));
    let maxSide = 0;
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (matrix[i][j] === '1') {
                if (i === 0 || j === 0) {
                    dp[i][j] = 1;
                } else {
                    dp[i][j] = Math.min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1;
                }
                maxSide = Math.max(maxSide, dp[i][j]);
            }
        }
    }
    return maxSide * maxSide;
}

console.log(maximalSquare([["1","0","1"],["1","1","1"],["1","1","1"]]));
console.log(maximalSquare([["0","1"],["1","0"]]));
